pkg/tsdb/elasticsearch/response_parser.go

```diff
@@ -56,19 +57,156 @@ func parseResponse(responses []*es.SearchResponse, targets []*Query) (*backend.Q
 
 		queryRes := backend.DataResponse{}
 
-		props := make(map[string]string)
-		err := processBuckets(res.Aggregations, target, &queryRes, props, 0)
-		if err != nil {
-			return &backend.QueryDataResponse{}, err
-		}
-		nameFields(queryRes, target)
-		trimDatapoints(queryRes, target)
+		if isDocumentQuery(target) {
+			err := processDocumentResponse(res, target, timeField, &queryRes)
+			if err != nil {
+				return &backend.QueryDataResponse{}, err
+			}
+			result.Responses[target.RefID] = queryRes
+		} else {
+			props := make(map[string]string)
+			err := processBuckets(res.Aggregations, target, &queryRes, props, 0)
+			if err != nil {
+				return &backend.QueryDataResponse{}, err
+			}
+			nameFields(queryRes, target)
+			trimDatapoints(queryRes, target)
 
-		result.Responses[target.RefID] = queryRes
+			result.Responses[target.RefID] = queryRes
+		}
 	}
 	return &result, nil
 }
 
+func processDocumentResponse(res *es.SearchResponse, target *Query, timeField string, queryRes *backend.DataResponse) error {
+	docs := make([]map[string]interface{}, len(res.Hits.Hits))
+	propNames := make([]string, 0)
+
+	for hitIdx, hit := range res.Hits.Hits {
+		var flattened map[string]interface{}
+		if hit["_source"] != nil {
+			flattened = flatten(hit["_source"].(map[string]interface{}))
+		}
+
+		doc := map[string]interface{}{
+			"_id":       hit["_id"],
+			"_type":     hit["_type"],
+			"_index":    hit["_index"],
+			"sort":      hit["sort"],
+			"highlight": hit["highlight"],
+			"_source":   flattened,
+		}
+
+		for k, v := range flattened {
+			doc[k] = v
+		}
+
+		for key := range doc {
+			// We want to add only unique propNames
+			if isStringInSlice(key, propNames) {
+				continue
+			} else {
+				propNames = append(propNames, key)
+			}
+		}
```

Comment:
I agree that sorting function (which should be quite simple, but I didn't come up with simple looking code for this in Go) is 👎 , but it makes other things MUCH more simple. If we went with your approach, we would have to:
- remove timeField from propNames to not have timefield duplicated
- because we pre-allocate resources for `allFields`, we would have to be checking when looping over propNames, if it should be set to `index` or `index + 1` based on if we have timefield. 
- If we decided to move `_source` to the last place, we could do it within this function again. 

I still think that using this function makes everything else more simple.

Reply:
sure, i see your point 👍